i6_models/primitives/mixup.py

```diff
@@ -0,0 +1,153 @@
+__all__ = ["MixupConfig", "Mixup"]
+
+import random
+from dataclasses import dataclass
+
+import torch
+
+from i6_models.config import ModelConfiguration
+
+
+@dataclass
+class MixupConfig(ModelConfiguration):
+    """
+    Attributes:
+        buffer_size: number of frames.
+        apply_prob: probability to apply mixup at all
+        lambda_min: minimum lambda value
+        lambda_max: maximum lambda value
+        max_num_mix: maximum number of mixups (random int in [1, max_num_mix])
+    """
+
+    buffer_size: int = 1_000_000
+    apply_prob: float = 1.0
+    max_num_mix: int = 4
+    lambda_min: float = 0.003
+    lambda_max: float = 0.3
+
+
+class FeatureBuffer:
+    """
+    The FeatureBuffer saves the feature from previous timeframes
+    after the buffer is full, the buffer will not be updated
+    """
+
+    def __init__(self, buffer_size: int, feature_dim: int):
+        self.filled = False
+        self.pos = 0
+        self.buffer_size = buffer_size
+        self.cache = torch.nn.parameter.Parameter(
+            data=torch.zeros((self.buffer_size, feature_dim)), requires_grad=False
+        )
+
+    def append(self, tensor: torch.Tensor):
```

Comment:
I would instead of filling the buffer and then stopping to add data, start overwriting again from the start of the buffer.
e.g. set `self.pos=0` again (but keep `self.filled=True`)

Reply:
Yes. In my original implementation, and also in Zoltans implementation, this is also the case. Why did you change this?

Reply:
The current implementation of `self.pos` is exactly the same as Zoltan's and I think it is right.
```
        delta_pos = min(self.buffer_size - self.pos, t_dim)
        end_pos = self.pos + delta_pos
        self.cache[self.pos : end_pos] = tensor[:delta_pos]
        self.pos = end_pos

        if end_pos == self.buffer_size:
            self.filled = True
            end_pos = t_dim - delta_pos
            self.cache[:end_pos] = tensor[delta_pos:]
            self.pos = end_pos
```
After `self.pos `reaches `self.buffer_size`, if we want to add new values, here `delta_pos` is 0, `delta_pos` is  `self.pos`. The if condition is right, `end_pos` is `t_dim` and the `self.pos` becomes `t_dim`(which is the length of the newly added values)

Reply:
The current implementation is technically correct and would work as-is.

I agree to 
> After self.pos reaches self.buffer_size, if we want to add new values, here delta_pos is 0

and then `end_pos` is `self.pos` so that we execute
```
self.cache[self.buffer_size: self.buffer_size] = tensor[:0]
```
i.e. no update is performed.

But what we would which to do is to change the current behavior and instead reset `self.pos=0` after we reach `self.buffer_size` so that the buffer is again overwritten from the start. This way, over the course of the training, we will not repeat the same mixup-noise over and over again but get fresh noise samples.

Reply:
no, for the current implementation, `delta_pos` is 0, so `end_pos` is `t_dim`, then the cache will be updated 
`self.cache[:t_dim] = tensor[0:]`

Reply:
ah yes, sorry, I see now how in the if clause in lines 57ff there would indeed be an update and the buffer will be overwritten from the start.